src/main/java/org/testng/TestClass.java

```diff
@@ -9,12 +9,14 @@
 import org.testng.xml.XmlTest;
 
 import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
 
 /**
  * This class represents a test class: - The test methods - The configuration methods (test and
  * method) - The class file
  */
-class TestClass extends NoOpTestClass implements ITestClass {
+class TestClass extends NoOpTestClass implements ITestClass, ITestClassConfigInfo {
```

Comment:
What if we use composition instead of inheritance?

Reply:
Could you please specify which part  you want to use composition, this code change only have add an interface.

Reply:
I don't know if the idea is good but as we talked about a test instance notion and when I see how the new API is used https://github.com/cbeust/testng/pull/2439#discussion_r546569701 then I think we can have an intermediate solution:

```java
interface ITestClassConfigInfo {
 XxxInstance getInstance(String id);
}
```

WithXxxInstance (I don't have a good name yet)
```java
class XxxInstance {
  List<ITestNGMethod> getBeforeClassMethods() {
    ...  
  }
}
```
@krmahadevan WDYT?

Reply:
@juherr - I am a bit confused here as well. Given the fact that `TestClass` is a package visible class, why do we need to add more complexity by leveraging composition ? (This new class could perhaps need more stuff just to get the configuration methods back to the caller). Wouldn't it be a lot more simpler for us to just have `TestClass` implement the new interface. Since this class is package visible, we can work on extracting out the object instantiation capabilities from it and start thinking of permanently moving them over to `TestClassInstance` (the new place?

If that doesn't sound quite right, then maybe I am still not understanding the proposition that you have put forth via the new class. So please help me understand it as well.

Reply:
@krmahadevan It's often a bit confused in my head too 😄 Your interrogations are always helping a lot! 👍 

About composition vs inheritance: I don't think one is more complex than the other. I think it is more a question of habits and the composition is helping to reduce design errors most of the time.

BTW, it was a proposition by anticipation because I supposed the complete PR will add more than only one method: 
* Is the issue exist for the other configuration methods? 
* If not, why will the workflow be different between configuration methods?

Then, from the design point of view, we are trying to store configuration methods for TestClass instances and it will be better to have a design that follows this semantic (`getInstance` -> `getBeforeClassMethods`).

Finally, it is the perfect opportunity to add one of the missing pieces of the model (`TestClassInstance` sounds good).